### ISSUE TITLE
Mallets no sound under Tubular Bells

### ISSUE THREAD

**Issue by jackokring** [LMMS/lmms#2455]
The first time a selection is made of a setting lower than tubular bells, a sound similar to a mangled buffer is played, then nothing ... until one of the working ones is selected. Further more the mangle is heard just once, and never upon reselection of a non working variety.

**Edited @tresf**

**Comment by Umcaruje** [comment]
LMMS 1.0.3 is an outdated version and its not supported. Please check if this bug exists on 1.1.3, which is the latest stable version. If you expirience the bug on 1.1.3, please offer steps to reproduce and your OS information.

You can get the latest version over at https://lmms.io/download/

**Comment by jackokring** [comment]
Will check when I get the master branch to compile. Currently cmake errors. Debian 8.2 latest apt-get toolchain, forked repo of master.

**Comment by jackokring** [comment]
git checkout stable-1.1
 ...
cmake .. -DWANT_VST=no
make

So I have 1.1.3 still no mallet sound beyond tubular bells

**Comment by tresf** [comment]
Reminds me slightly of this https://github.com/LMMS/lmms/pull/1529.

@jackokring if you wrap your command/codeblocks in \`\`\`bash [stuff] \`\`\` tags, they'll display nicely.

```bash
echo "foo"
```

**Comment by jackokring** [comment]
OK, minor update, on inspection (listening) of the sound not output, the first time a selection is made of a setting lower than tubular bells, a sound similar to a mangled buffer is played, then nothing ... until one of the working ones is selected. Further more the mangle is heard just once, and never upon reselection of a non working variety.

**Comment by Umcaruje** [comment]
Reproduced on elementary OS Freya on the master branch. 

None of the settings lower than Tubular Bells produce sound, but when I check the 'Bowed' checkbox, they start making a glitched sound, as if they are stuttered ...

**Comment by jackokring** [comment]
I check the box and the sound worked, so seems to me the box is implicitly ticked and should be removed from the GUI.

**Comment by tresf** [comment]
@jackokring can you please retitle and fix the original description?

**Comment by zonkmachine** [comment]
So it's the bowed instruments then. It could be the stk package.

**Comment by zonkmachine** [comment]
Yeah, this is in 0.4.14 too.

@musikBear You reported on this issue on the mailing list around the 1.0.3 release.
What version of stk do you have?

**Comment by tresf** [comment]
He'll have whatever was bundled at the time of `1.1.3`/mingw32

**Comment by zonkmachine** [comment]
OK. From the release notes for stk 4.4.3
`Updates to Bowed class (thanks to Esteban Maestre!)`

This place us in time around Ubuntu 12.04 and lmms-0.4.10

**Comment by zonkmachine** [comment]
lmms-0.4.2 / stk-4.2.0 Same issue, so the changes in stk above had nothing to do with it.

**Comment by musikBear** [comment]
@zonkmachine / 
current Systems Tool Kit c++ is the one @tresf bundled in the test RC g23d2824. This RC is '100%' stable for me. The new edit-features in piano-roll are incredible useful (saying that not because i suggested those :p, but current stk is ofcause of no relevance to the one at 1.1.3 release time. If i Can make any kind of experiment, that can be of interests? just let me know

**Comment by zonkmachine** [comment]
> If i Can make any kind of experiment, that can be of interests? just let me know

Yes. Just test mallets and see if you find anything of interest.

**Comment by musikBear** [comment]
> Yes. Just test mallets and see if you find anything of interest.

@zonkmachine 
Made a testbed with escalating automation for all instrument-dials.
I used wave-recording to illustrate the various functionalities. 
The first series are for the presets:
*Marimba, vibraphone, agogo, wood1, reso, wood2, beats, two-fixed, clump*
That all have the same controller-dials
**Hardness**
![malletsvibraphonehardnessauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549250/0e1d6f2a-e302-11e5-9271-19b5b3d10260.png)
The dial does not change anything. doubt it connected.

**Position**
![malletsvibraphoneposauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549279/739f9b70-e302-11e5-9b93-8909abd1acf9.png)

**Stick mix**
![malletsvibraphonestickmixauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549288/8aee123e-e302-11e5-9de2-feeb53bfcd35.png)

**Vib Gain**
![malletsvibraphonevibgainauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549295/a4f20816-e302-11e5-8d5e-938caf7d7f92.png)
Looks fine

**Vib freq**
![malletsvibraphonevibfreqauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549303/c19d5916-e302-11e5-9bed-a58a11cadc96.png)
Looks fine

**Spread**
Waveform not influenced, but it is ok.

Second series
Preset *Tubular bells*
**Modulation**
![malletstubularmodauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549365/c7bb3bb4-e303-11e5-8476-3ae3ed2ed918.png)

**Crossfade**
![malletstubularcrossfadeauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549372/e9eadcb2-e303-11e5-8f9c-a8d18f1f06f5.png)

**ADSR**
![malletstubularadsrauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549377/12fcd718-e304-11e5-9c19-0f1fa1b10559.png)
Looks fine

**LFO speed**
![malletstubularlfospeedauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549397/56d24f86-e304-11e5-9ffe-c7d763f931f0.png)
Looks fine

@zonkmachine 
Made a testbed with escalating automation for all instrument-dials.
I used wave-recording to illustrate the various functionalities. 
The first series are for the presets:
*Marimba, vibraphone, agogo, wood1, reso, wood2, beats, two-fixed, clump*
That all have the same controller-dials
**Hardness**
![malletsvibraphonehardnessauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549250/0e1d6f2a-e302-11e5-9271-19b5b3d10260.png)
The dial does not change anything. doubt it connected.

**Position**
![malletsvibraphoneposauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549279/739f9b70-e302-11e5-9b93-8909abd1acf9.png)

**Stick mix**
![malletsvibraphonestickmixauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549288/8aee123e-e302-11e5-9de2-feeb53bfcd35.png)

**Vib Gain**
![malletsvibraphonevibgainauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549295/a4f20816-e302-11e5-8d5e-938caf7d7f92.png)
Looks fine

**Vib freq**
![malletsvibraphonevibfreqauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549303/c19d5916-e302-11e5-9bed-a58a11cadc96.png)
Looks fine

**Spread**
Waveform not influenced, but it is ok for **all**.

Second series
Preset *Tubular bells*
**Modulation**
![malletstubularmodauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549365/c7bb3bb4-e303-11e5-8476-3ae3ed2ed918.png)
The dial does not change anything. doubt it connected.


**Crossfade**
![malletstubularcrossfadeauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549372/e9eadcb2-e303-11e5-8f9c-a8d18f1f06f5.png)
Looks fine

**ADSR**
![malletstubularadsrauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549377/12fcd718-e304-11e5-9c19-0f1fa1b10559.png)
Looks fine

**LFO Depth**
![malletstubularlfodepthauto16bar](https://cloud.githubusercontent.com/assets/6662066/13549403/6ba905a8-e304-11e5-95de-16367b3656f9.png)
Looks fine

3. series
Presets
*Uniform-bar, Tuned-bar, glass, Tibetan-bowl*
These 4 presets is not working at all.
They produce no sound unless *bowed*, but it is only 'noise'. In the case of *Uniform-bar* **extremely** loud, and dangerous for ears/speakers. ! Be careful !

**Comment by zonkmachine** [comment]
@musikBear We got pretty much the same result then. I'll add to this that:

 * The Vibrato knob in the last set of instruments isn't fully implemented and therefore 'stuck' in one position.
 * On reload the default/first set of knobs will be shown even if you have a sound after and including Tubular Bells. You need to manually change to another instrument and then back to what you had in order for the knobs to work.
 * Vib freq only works for the Vibraphone.

* On the last set, Uniform - Tibetan, the Pressure and Speed knobs really works but they go a bit glitchy if you set them to 0.0 . I'll fix this by setting the lower limit to 0.1. fixed.

**Comment by musikBear** [comment]
:+1: very good

**Comment by zonkmachine** [comment]
@LMMS/developers
What I'm thinking about this right now is:
 - Leave Mallets untouched.
 - Introduce 'Mallets 2' which has only the Modalbar stuff( first nine instruments ) but fixed up so the controls work properly. #2670 
 - Tubular Bells is taken over by Borg Jr. #2668
 - Don't do anything new with BandedWG ( Last four instruments, Uniform bar - Tibetan bowl ). It will take unnecessary development time to get it right for something that most likely won't be used that much. You can already experiment with this stuff in Mallets. Just check 'bowed' and add envelopes, volume/cutoff. Knobs 'Motion' and 'Vibrato' doesn't work but 'Pressure' and 'Speed' does.
It's glitchy and not in a good way. Use a limiter...

This way you end up with two new instruments that works and has more manageable code.

**Comment by Fastigium** [comment]
> An updated Mallets will open older projects wrong because people will have turned the knobs

Maybe that could be addressed using `DataFile::upgrade()`?

**Comment by Umcaruje** [comment]
>Maybe that could be addressed using DataFile::upgrade()?

^What Fastigium said, making another version of the same plugin is the wrong approach to this problem imo.

**Comment by zonkmachine** [comment]
> Maybe that could be addressed using DataFile::upgrade()?

I'm sure there's way to manage the update but we need to take care of the other bugs first and they're in code related to/or in, the BandedWG module.
https://github.com/LMMS/lmms/pull/2671#issuecomment-195940322
If you find what's wrong in there bump me about it but for now I'm stuck. I just think the most clever way is to roll along with the two spin-offs of Mallets and instead find a way to hide old stuff in the menu and only launch them in project's that already got them and for users who choose to activate all plug ins.

**Comment by Fastigium** [comment]
> If you find what's wrong in there bump me about it but for now I'm stuck

My plate is full with crash-fixing for the moment, but perhaps you might gain some insight by using conditional breakpoints in gdb. They should allow you to halt execution when a spike happens and trace where it might come from (`bt full` will show you values of local variables). Hope that helps, otherwise I'll see if I can take a peek sometime soon.

**Comment by zonkmachine** [comment]
OK. I'll try and move my upgrades to Mallets instead and will read up a bit on gdb.

**Comment by Fastigium** [comment]
You may also find gdb's `frame` command useful. It allows you to enter the scope of any function in the backtrace and print members of local objects and such. Was invaluable to me in fixing those "index out of range" crashes.

**Comment by zonkmachine** [comment]
Mallets is now ready for testing. https://github.com/LMMS/lmms/pull/2670

I used the same backward compatibility method as Kicker as I recently worked with it and it's familiar.
Most bugs fixed. Mallets seem more to have been abandoned before really completed and left with more or less the preset values from stk modalbar, BandedWG presets turned default off and only Tubular Bells from stk::FM really working full out.

> They should allow you to halt execution when a spike happens

I haven't been able to reproduce the spikes.
